### PR TITLE
DM-37555: Update pipeline documentation for image differencing

### PR DESCRIPTION
**Pull request by BrunoSanchez**
Big documentation update, removed old docs and added basic help info.

### CODE REVIEW

**Comment by parejkoj** [comment]
Please fix the PR title. See here:
https://developer.lsst.io/work/flow.html#make-a-pull-request

File: doc/lsst.ip.diffim/index.rst [hunk 1]
```diff
@@ -15,14 +34,72 @@ Using lsst.ip.diffim
...
+
+Single frame processing
+-----------------------
+
+Single image frame usage can be carried out by importing the ``lsst.ip.diffim`` as a library of tasks, as the following example shows:
+
+.. code-block:: python
+
+   from lsst.ip.diffim import AlardLuptonSubtractConfig 
+   from lsst.ip.diffim import AlardLuptonSubtractTask
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/index.rst:80]: Instead I'd do `from lsst.ip.diffim import subtractImages`; then you can use that namespace below.

File: doc/lsst.ip.diffim/overview_ipdiffim.rst [hunk 2]
```diff
@@ -0,0 +1,107 @@
+################
+Package overview
+################
+
+The package has different types of pipeline tasks that can perform image 
+subtraction and transient source detection. In some cases there are more
+than one specific implementation available per type of task, this is, 
+different options for image subtraction and source detection.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/overview_ipdiffim.rst:8]:
```suggestion
``ip_diffim`` contains pipeline tasks to perform image subtraction and transient source detection on calibrated exposures and template images.
The various implementations of these tasks are described below.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureScoreTask.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+Description
+===========
+
+This task detects sources in score-like difference images, such as the ones obtained with pre-convolution
+image differencing, or ZOGY "S" images.
+This means that cross-convolution is not needed, and the task obtains maximum-likelihood source candidates
+in a single step. 
+Once candidate sources are obtained this task performs the same operations as DetectAndMeasureTask, 
+for further reference consult the documentation of DetectAndMeasureTask.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureScoreTask.rst:19]: Please link to said tasks when mentioned in text. Do this for the other files above, too, please.

```suggestion
Once candidate sources are obtained this task performs the same operations as `~lsst.ip.diffim.DetectAndMeasureTask`;
for further reference consult the documentation of `~lsst.ip.diffim.DetectAndMeasureTask`.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+##################################
+
+AlardLuptonPreconvolveSubtractTask is a modification of AlardLuptonSubtractTask behavior, and it performs image differencing between the `template` image and a custom kernel cross-correlated `science` image.
+
+.. _lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask-description:
+
+Description
+===========
+
+This task takes the pair of science and template exposures, together with the science image source catalog, and produces a difference and matched template exposures as outputs.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst:14]:
```suggestion
This task takes the pair of science and template exposures, together with the science image source catalog, and produces difference and matched template exposures as outputs.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+The process involves selecting sources from the science exposure source catalog, obtaining the convolution kernel, applying this to the `target` image and obtaining the subtraction.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst:16]:
```suggestion
The process involves selecting sources from the science exposure source catalog, obtaining the convolution kernel, applying this to the `target` image, and subtracting the images.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+As discussed in the general package overview the task has two possible modes: one convolving the `template` image and the alternative convolving the `science` image.
+Both modes share the preparation sub-tasks, but the method that performs the convolution is not the same.
+The parameter that controls this behavior is ``mode`` (check ``lsst.ip.diffim.AlardLuptonSubtractConfig``), and it allows ``convolveScience``, ``convolveTemplate`` and finally ``auto`` which means that the decision is made at runtime automatically.
+
+Finally, mask planes are correctly set for the difference image and matched template exposure outputs.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst:22]:
```suggestion
Finally, mask planes are set on the difference and matched template outputs based on the masks of the input science and template images.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonSubtractTask.rst [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+AlardLuptonSubtractTask performs image differencing, including decorrelation.
+
+
+.. _lsst.ip.diffim.AlardLuptonSubtractTask-description:
+
+Description
+===========
+
+This task takes the pair of science and template exposures, together with the science image source catalog, and produces a difference and matched template exposures as outputs.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonSubtractTask.rst:15]:
```suggestion
This task takes the pair of science and template exposures, together with the science image source catalog, and produces difference and matched template exposures as outputs.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonSubtractTask.rst [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+The process involves selecting sources from the science exposure source catalog, obtaining the convolution kernel, applying this to the `target` image and obtaining the subtraction.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonSubtractTask.rst:17]:
```suggestion
The process involves selecting sources from the science exposure source catalog, obtaining the convolution kernel, applying this to the `target` image, and subtracting the images.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonSubtractTask.rst [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+As discussed in the general package overview the task has two possible modes: one convolving the `template` image and the alternative convolving the `science` image.
+Both modes share the preparation sub-tasks, but the method that performs the convolution is not the same.
+The parameter that controls this behavior is ``mode`` (check ``lsst.ip.diffim.AlardLuptonSubtractConfig``), and it allows ``convolveScience``, ``convolveTemplate`` and finally ``auto`` which means that the decision is made at runtime automatically.
+
+Finally, mask planes are correctly set for the difference image and matched template exposure outputs.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonSubtractTask.rst:23]:
```suggestion
Finally, mask planes are set on the difference and matched template outputs based on the masks of the input science and template images.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureTask.rst [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+DetectAndMeasureTask performs source detection and execution of measurement plugins in difference images.
+
+.. _lsst.ip.diffim.DetectAndMeasureTask-description:
+
+Description
+===========
+
+This task cross-convolves the difference image with a PSF approximation to obtain maximum-likelihood source candidates.
+Measurements are run on all the detections, including positive flux and negative flux candidates.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureTask.rst:15]:
```suggestion
Measurements are run on all the detections, including positive and negative flux candidates.
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureTask.rst [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+Near sources will be combined, enabling trailed sources as well as dipole-like source detection.
```
Comment (parejkoj) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureTask.rst:16]:
```suggestion
Close sources will be combined, enabling trailed sources as well as dipole-like source detection.
```

**Review by parejkoj — CHANGES_REQUESTED** [review]
One suggestion: please use the "one sentence per line" format described here in the overview:

https://developer.lsst.io/restructuredtext/style.html#text-wrapping

instead of breaking your lines at ~80chars. It makes future modifications easier, and makes parsing the raw ReST easier. I'll look at the overview file after it's refactored, so it's easier to comment on.

**Comment by BrunoSanchez** [comment]
OK, all changes implemented.
Sorry I took long to realize this was reviewed already.

File: doc/lsst.ip.diffim/index.rst [hunk 12]
```diff
@@ -15,14 +34,47 @@ Using lsst.ip.diffim
...
-.. toctree::
+.. toctree
    :maxdepth: 1
 
-   code_notes
-   AL_implementation
+Production pipelines
+--------------------
+
+This package is generally being used in production pipelines like ``lsst.ap.pipe`` and ``lsst.drp.pipe``, designed for processing large collections of survey data.
```
Comment (isullivan) [inline doc/lsst.ip.diffim/index.rst:46]:
```suggestion
This package is used in production pipelines like ``lsst.ap.pipe`` and ``lsst.drp.pipe``, designed for processing large collections of survey data.
```

File: doc/lsst.ip.diffim/overview_ipdiffim.rst [hunk 13]
```diff
@@ -0,0 +1,78 @@
+################
+Package overview
+################
+
+``ip_diffim`` contains pipeline tasks to perform image subtraction and transient source detection on calibrated exposures and template images.
```
Comment (isullivan) [inline doc/lsst.ip.diffim/overview_ipdiffim.rst:5]:
```suggestion
``ip_diffim`` contains pipeline tasks to perform image subtraction and transient source detection on calibrated exposures, template images, and their image differences.
```

File: doc/lsst.ip.diffim/overview_ipdiffim.rst [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+The various implementations of these tasks are described below.
+
+Image subtraction tasks
+=======================
+
+The image differencing tasks in this package are implementations of the Alard & Lupton (1998) algorithm (AL) [AL_1998]_, and this can be summarised as applying a kernel determination algorithm in order to match observational properties between two images. 
+The images equalized vary according to which task is used, in combination with configuration parameters of such tasks.
```
Comment (isullivan) [inline doc/lsst.ip.diffim/overview_ipdiffim.rst:12]:
```suggestion
The method to match images varies according to which task is used, in combination with configuration parameters of such tasks.
```

File: doc/lsst.ip.diffim/overview_ipdiffim.rst [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+
+AlardLuptonPreconvolveSubtractTask
+----------------------------------
+
+The `pre-convolve` task runs a preparation stage before applying the AL algorithm itself, so the determination of the kernel for the `template` image is never a "deconvolution" kernel (negative wing profile, as shown in [Kovacs_DMTN-179]_).
+
+In this stage the `science` image is transformed with a convolution kernel :math:`H`, that is tipically the PSF of this image, becoming a "score image" for detection. 
+Once this transformation is applied, standard A&L algorithm derives a kernel for the `template` image so it matches the `science score` image properties, and the subtraction is obtained. \
+It is needed to clarify that this result is again a "score" image, and Maximu-Likelihood source detection can be run on it without the standard PSF cross-correlation step.
```
Comment (isullivan) [inline doc/lsst.ip.diffim/overview_ipdiffim.rst:37]:
```suggestion
It is needed to clarify that this result is again a "score" image, and Maximum-Likelihood source detection can be run on it without the standard PSF cross-correlation step. 
```

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+This task takes the pair of science and template exposures, together with the science image source catalog, and produces difference and matched template exposures as outputs.
+
+The process involves selecting sources from the science exposure source catalog, obtaining the convolution kernel, applying this to the `target` image, and subtracting the images.
+
+As discussed in the general package overview the task has two possible modes: one convolving the `template` image and the alternative convolving the `science` image.
+Both modes share the preparation sub-tasks, but the method that performs the convolution is not the same.
+The parameter that controls this behavior is ``mode`` (check `~lsst.ip.diffim.AlardLuptonSubtractConfig`), and it allows ``convolveScience``, ``convolveTemplate`` and finally ``auto`` which means that the decision is made at runtime automatically.
+
+Finally, mask planes are set on the difference and matched template outputs based on the masks of the input science and template images.
```
Comment (isullivan) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst:22]: This description doesn't sound right for the preconvolution task. It should state that it creates a Score image, and we don't use mode.

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureScoreTask.rst [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+DetectAndMeasureScoreTask performs source detection and execution of measurement plugins in difference images.
+
+.. _lsst.ip.diffim.DetectAndMeasureScoreTask-description:
+
+Description
+===========
+
+This task detects sources in score-like difference images, such as the ones obtained with pre-convolution image differencing, or ZOGY "S" images.
+This means that cross-convolution is not needed, and the task obtains maximum-likelihood source candidates in a single step. 
+Once candidate sources are obtained this task performs the same operations as `~lsst.ip.diffim.DetectAndMeasureTask`; for further reference consult the documentation of `~lsst.ip.diffim.DetectAndMeasureTask`.
```
Comment (isullivan) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.DetectAndMeasureScoreTask.rst:16]: Add a note here that the diaSource *measurement* is still performed on the difference image, not the Score image.

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.MakeKernelTask.rst [hunk 18]
```diff
@@ -11,51 +11,28 @@ MakeKernelTask creates a PSF-matching kernel for two images.
...
-called: ``BuildSingleKernelVisitor``, ``KernelSumVisitor``, ``BuildSpatialKernelVisitor``,
-and ``AssessSpatialKernelVisitor``.
-
-Upon initialization, the kernel configuration is defined by
-``self.config.kernel.active``. The task creates an `~lsst.afw.math.Warper` from the
-subConfig ``self.config.kernel.active.warpingConfig``. A schema for the selection
-and measurement of candidate `~lsst.ip.diffim.KernelCandidates` is defined, and
-used to initize subTasks selectDetection (for candidate detection) and
-selectMeasurement(for candidate measurement).
+Build a Psf-matching kernel using two input images, either as MaskedImages (in which case they need to be astrometrically aligned) or Exposures (in which case astrometric alignment will happen by default but may be turned off).
```
Comment (isullivan) [inline doc/lsst.ip.diffim/tasks/lsst.ip.diffim.MakeKernelTask.rst:14]: I don't believe this Task can handle Exposures that are not astrometrically aligned any more. It was certainly my intention to remove that functionality, so if it is still there we should disable it.

**Review by isullivan — APPROVED** [review]
Looks good! I had a few wording changes.

**Review by parejkoj — APPROVED** [review]

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst (hunk 16 above)
Comment (BrunoSanchez) [reply doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst:22]: Wow, that's a mistake! Thanks!

Comment (BrunoSanchez) [reply doc/lsst.ip.diffim/tasks/lsst.ip.diffim.AlardLuptonPreconvolveSubtractTask.rst:22]: Well, I think I somehow overwrote the changes that I made for this page. I did write a good explanation. I am afraid it is lost.

File: doc/lsst.ip.diffim/tasks/lsst.ip.diffim.MakeKernelTask.rst (hunk 18 above)
Comment (BrunoSanchez) [reply doc/lsst.ip.diffim/tasks/lsst.ip.diffim.MakeKernelTask.rst]: I took a look and it seems this functionality is not there anymore. I couldn't find it at least. Changed the docs to reflect this.

**Pull request merged**